yorkie/src/main/kotlin/dev/yorkie/util/SplayTreeSet.kt

```diff
@@ -0,0 +1,364 @@
+package dev.yorkie.util
+
+import com.google.common.annotations.VisibleForTesting
+
+/**
+ * SplayTreeSet is weighted binary search tree set based on Splay tree.
+ * original paper on Splay Trees:
+ * @link https://www.cs.cmu.edu/~sleator/papers/self-adjusting.pdf
+ */
+// should SplayTree implement MutableSet?
+internal class SplayTreeSet<V>(private val lengthCalculator: (V) -> Int = { 1 }) {
+    private val valueToNodes = mutableMapOf<V, Node<V>>()
+
+    @VisibleForTesting
+    var root: Node<V>? = null
+        private set
+
+    val length
+        get() = root?.weight ?: 0
+
+    private val maximum: Node<V>?
+        get() {
+            var node = root
+            while (node?.right != null) {
+                node = node.right
+            }
+            return node
+        }
+
+    private fun subtree(node: Node<V>?): SplayTreeSet<V> {
+        return SplayTreeSet(lengthCalculator).also {
+            it.root = node
+        }
+    }
+
+    /**
+     * returns the Value and offset of the given index.
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun find(pos: Int): ValueToOffset<V> {
+        var target = pos
+        val root = root
+        if (root == null || target < 0) {
+            return ValueToOffset.Empty as ValueToOffset<V>
+        }
+        var node: Node<V> = root
+        while (true) {
+            if (target <= node.leftWeight) {
+                node = requireNotNull(node.left)
+            } else if (node.right != null && node.leftWeight + node.length < target) {
+                target -= node.leftWeight + node.length
+                node = requireNotNull(node.right)
+            } else {
+                target -= node.leftWeight
+                break
+            }
+        }
+        if (target > node.length) {
+            error("out of index range: pos: $target > node.length: ${node.length}")
```

Comment:
I wonder how Android-based SDKs handle errors.

1. Use `error` not defining things like custom errors?
2. How do we receive and handle this error on the caller?

Reply:
1. We can use custom errors or exceptions provided by Java & Kotlin.

in fact, `error` is a helper method provided by Kotlin which throws `IllegalStateException`.
Which I thought would be a good choice at the time, but I think it should be changed to throw `IndexOutOfBoundsException` :)

2. we can simply `try-catch`. But Kotlin does not have checked exceptions (https://kotlinlang.org/docs/exceptions.html#checked-exceptions)